File: src/FarmEngine.Core/Replay.cs
using System.Text.Json.Serialization;
using FarmEngine.Schemas;

namespace FarmEngine.Core;

/// <summary>
/// One entry of a replay input log. JSON shape matches the TS union:
/// <c>{"kind":"command","command":{…}}</c> / <c>{"kind":"tick","ticks":N}</c>.
/// </summary>
[JsonPolymorphic(TypeDiscriminatorPropertyName = "kind")]
[JsonDerivedType(typeof(CommandInput), "command")]
[JsonDerivedType(typeof(TickInput), "tick")]
public abstract record ReplayInput
{
    [JsonIgnore] public abstract string Kind { get; }
}

public sealed record CommandInput(Command Command) : ReplayInput { [JsonIgnore] public override string Kind => "command"; }

public sealed record TickInput(double Ticks) : ReplayInput { [JsonIgnore] public override string Kind => "tick"; }

public sealed record ReplayResult(GameState State, string Hash, List<Effect> Effects);

/// <summary>
/// Replay harness (port of replay.ts): run a scripted input log through the
/// engine and hash the final state. Golden replay tests fail on any
/// nondeterminism.
/// </summary>
public static class Replay
{
    public static ReplayResult RunReplay(EngineContext ctx, GameState initialState, IEnumerable<ReplayInput> inputs)
    {
        var state = initialState;
        var effects = new List<Effect>();

        foreach (var input in inputs)
        {
            if (input is TickInput tick)
            {
                var step = Engine.AdvanceTick(ctx, state, tick.Ticks);
                state = step.State;
                effects.AddRange(step.Effects);
            }
            else if (input is CommandInput command)
            {
                var step = Engine.ApplyCommand(ctx, state, command.Command);
                state = step.State;
                effects.AddRange(step.Effects);
            }
        }

        return new ReplayResult(state, Hash.HashState(state), effects);
    }

    /// <summary>TS <c>command(cmd)</c>.</summary>
    public static ReplayInput Cmd(Command command) => new CommandInput(command);

    /// <summary>TS <c>command(cmd)</c> under its PascalCase name (alias of <see cref="Cmd"/>).</summary>
    public static ReplayInput Command(Command command) => new CommandInput(command);

    /// <summary>TS <c>ticks(count)</c>.</summary>
    public static ReplayInput Ticks(double count) => new TickInput(count);
}

File: crates/farm-sim/src/replay.rs
//! Replay harness (port of `Replay.cs` / replay.ts): run a scripted input log through the engine
//! and hash the final state. Golden replay tests fail on any nondeterminism.

use crate::commands::Command;
use crate::engine;
use crate::engine_types::{Effects, EngineContext};
use crate::hash;
use crate::schema::GameState;
use serde::{Deserialize, Serialize};

/// One entry of a replay input log. JSON shape matches the TS union:
/// `{"kind":"command","command":{…}}` / `{"kind":"tick","ticks":N}`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum ReplayInput {
    Command { command: Command },
    Tick { ticks: f64 },
}

/// TS `ReplayResult`: the final state hash and every effect, in order (the state itself was
/// updated in place).
#[derive(Debug, Clone, PartialEq, Default)]
pub struct ReplayResult {
    pub hash: String,
    pub effects: Effects,
}

pub fn run_replay(ctx: &EngineContext, state: &mut GameState, inputs: &[ReplayInput]) -> ReplayResult {
    let mut effects = Vec::new();
    for input in inputs {
        match input {
            ReplayInput::Tick { ticks } => effects.extend(engine::advance_tick(ctx, state, *ticks)),
            ReplayInput::Command { command } => effects.extend(engine::apply_command(ctx, state, command)),
        }
    }
    ReplayResult { hash: hash::hash_state(state), effects }
}

/// TS `command(cmd)`.
pub fn command(command: Command) -> ReplayInput {
    ReplayInput::Command { command }
}

/// TS `ticks(count)`.
pub fn ticks(count: f64) -> ReplayInput {
    ReplayInput::Tick { ticks: count }
}
